lisp/ess-r-mode.el

```diff
@@ -355,11 +348,9 @@ To be used as part of `font-lock-defaults' keywords."
      (ess-make-source-refd-command-function . #'ess-r-make-source-refd-command)
      (ess-format-eval-message-function      . #'ess-r-format-eval-message)
      (ess-install-library-function          . #'ess-r-install-library)
-     (ess-eldoc-function                    . #'ess-r-eldoc-function)
```

Comment:
Why these variables are removed? There are two main functions of customize alist. First is that we could move move variables across buffers. Second is that we avoid repetition when installing modes. I think both functions are not entirely obliterated with the derived mode system.

Reply:
I can't tell which variables you're talking about exactly. `ess-eldoc-function`? It's now set in the appropriate major modes. For R that's ess-r-mode and inferior-ess-r-mode. We repeat ourselves a bit but I think the code is much simpler: (If we're trying to separate ess-r-mode from ess-mode this will also have to be done at some point) 

this is what we remove:

```elisp
    ;; eldoc)
    (require 'eldoc)
    (when (and ess-eldoc-function ;; if mode provide this, it suports eldoc
               (or (and (not inferior) ess-use-eldoc)
                   (and inferior (eq ess-use-eldoc t))))
      (when (> eldoc-idle-delay 0.4) ;; default is too slow for paren help
        (set (make-local-variable 'eldoc-idle-delay) 0.1))
      (set (make-local-variable 'eldoc-documentation-function) ess-eldoc-function)
      (eldoc-mode 1))
``` 

in favor of the much simpler:

```elisp
  ;; eldoc
  (add-function :before-until (local 'eldoc-documentation-function)
                #'ess-r-eldoc-function)
  (when ess-use-eldoc (eldoc-mode))
```

Reply:
> I can't tell which variables you're talking about exactly. 

Everything that you have removed from `ess-r-customize-alist` -  `ess-ac-sources`, `ess-company-backends` 

> We repeat ourselves a bit but I think the code is much simpler:

Ok, looks like some variables are now bypassing ess-customize. This is inconsistent behavior. Either all or none.

>  ;; eldoc
  (add-function :before-until (local 'eldoc-documentation-function)

Why do we need to advice eldoc-documentation-function if we can just install our own eldoc-documetnation-function? 

Do you use advice in other contexts as well?

Reply:
We had to bypass some of the -customize-alist stuff because of the timing of when variables get set. If we've got `ess-r-mode` turned on in .R files from auto-mode-alist, the first thing it does is call `(ess-r-mode)` (well, technically not the *first*, but close enough). Using define-derived-mode means that they syntax table is set immediately from ess-r-mode-syntax-table, which bypasses the customize alists since they aren't set yet. We could of course go back and reset it but when I tried that I ran into problems (I forget exactly what). It seems simpler and more consistent with what other major modes do to set this up properly in the first place.

As complicated as this PR is overall, the resulting code is /less/ complicated, I think. It's at least shorter --- about  300 lines shorter.

Reply:
> add-function is a documented way of modifying -function variables.

Hmm. This is something new to me :| 

> Isn't it ok to do it progressively? 

I don't think this could work, but I cannot judge the impact of these changes. 

Let's define the goal. Do we want to get rid of the customize alist in the first place?

Good news is that we don't use customize alist for porting variables from one buffer to another that often. One has to go through all those `ess-setq-vars-local` and see case by case if it could be replaced with something else. If variable porting is no longer a concern it all comes to whether we want to copy paste same setq code across different non-derived modes.

Reply:
> Do we want to get rid of the customize alist in the first place?

I'd prefer to, eventually. I find them a little weird/confusing. We've got ess-customize-alist which isn't buffer local and gets changed and ess-local-customize-alist which is buffer-local. And then another customize-alist for each language. It's hard to keep track of what's happening when.

Most of the ess-setq-vars-local calls can be replaced with just calling the appropriate major mode, I think.

That said, I don't think we have to get rid of the customize-alists in this PR.

Reply:
Ok. I have the same feeling. It's better to use consistent and standard tooling, even if one has to write a few more lines of setqs here and there. We can also resort to some simpler tricks than ess-customize-alist for variable porting and for reducing copy-paste burden. 

So let's aim at removing customize alist by next release and also sticking to only standard inheritance mechanisms (ess-derived-mode and cl-generic). I would also like to remove ess-defgeneric. The purpose of the internal :overwrite could be achieved by an extra cl-generic per function, and it's not really a big deal given the extra benefits of the standard tooling like xref which works with it.

Reply:
OK, great, sounds like a plan to me. I'll open some issues so we don't lose track of what we talked about here. 

As for this PR, I *think* I've responded to everything, let me know if I missed something.

Reply:
I agree with everything said here. A main goal for the ESS rewrite is to switch towards standard Emacs tooling and idioms, where feasible.